src/test/java/org/influxdb/impl/RetryCapableBatchWriterTest.java

```diff
@@ -0,0 +1,89 @@
+package org.influxdb.impl;
+
+import org.influxdb.InfluxDB;
+import org.influxdb.InfluxDBException;
+import org.influxdb.dto.BatchPoints;
+import org.influxdb.dto.Point;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.runner.JUnitPlatform;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.*;
+
+@RunWith(JUnitPlatform.class)
+public class RetryCapableBatchWriterTest {
+
+  BatchPoints getBP(int count) {
+    BatchPoints.Builder b = BatchPoints.database("d1");
+    for (int i = 0; i < count; i++) {
+      b.point(Point.measurement("x1").addField("x", 1).build()).build();
+    }
+    return b.build();
+  }
+
+  @Test
+  public void test() {
+    InfluxDB mockInfluxDB = mock(InfluxDBImpl.class);
+    BiConsumer errorHandler = mock(BiConsumer.class);
+    RetryCapableBatchWriter rw = new RetryCapableBatchWriter(mockInfluxDB, errorHandler,
+            150, 100);
+    BatchPoints bp0 = getBP(5);
+    BatchPoints bp1 = getBP(90);
+    BatchPoints bp2 = getBP(90);
+    BatchPoints bp3 = getBP(8);
+    BatchPoints bp4 = getBP(100);
+
+    Exception nonRecoverable = InfluxDBException.buildExceptionForErrorState("database not found: cvfdgf");
+    Exception recoverable = InfluxDBException.buildExceptionForErrorState("cache-max-memory-size exceeded 104/1400");
+    Mockito.doThrow(nonRecoverable).when(mockInfluxDB).write(bp0);
+    Mockito.doThrow(recoverable).when(mockInfluxDB).write(bp1);
+    Mockito.doThrow(recoverable).when(mockInfluxDB).write(bp2);
+    Mockito.doThrow(recoverable).when(mockInfluxDB).write(bp3);
+    // first one will fail with non-recoverable error
+    rw.write(Collections.singletonList(bp0));
+    // second one will fail with recoverable error
+    rw.write(Collections.singletonList(bp1));
```

Comment:
Let's suppose that `bp1` is failing with recoverable error and `bp2` is writing successfully (i.e. WriteResultOutcome.WRITTEN).
If both `bp1` and `bp2` are writing to the same point in time (but `bp2` contains the most recent values), I think the current retry logic will write `bp1` and overwrite the existing values with old ones. Am I wrong?

Reply:
Just to be clear:
`bp1`
```
time                 sometag somevalue
----                 ------- ----
2017-06-20T15:32:46Z foo     ON
2017-06-20T15:32:47Z foo     OFF
2017-06-20T15:32:48Z foo     ON
```

`bp2`
```
time                 sometag somevalue
----                 ------- ----
2017-06-20T15:32:48Z foo     OFF
2017-06-20T15:32:49Z bar     OFF
2017-06-20T15:32:50Z bar     OFF
```
`2017-06-20T15:32:48Z foo OFF` is the most recent point but as we are retrying `bp1`, it will be overwritten with `2017-06-20T15:32:48Z foo ON`

Reply:
@fmachado No, under usual circumstances the implementation won't try to write bp2 after bp1 write failed on a recoverable error, it will just get queued in the RetryCapableBatchWriter. The case you describe might happen when the bufferLimit gets exceeded, in that case oldest batch points that failed on a recoverable error get forgotten.

This test is a junit test targetted on the RetryCapableBatchWriter internals - there are no writes into InfluxDB happening right now, that is why the conflicts in timestamps between bp1 and bp2 don't matter. Hoan is building BatchOptionsTest for which the implementation will be a black box.

Reply:
RetryCapableBatchWriter follows the approach of telegraf, which is aware of and guarantees the chronological order of the written batchpoints, even when bufferLimit is exceeded.
For example, with Fernando's case, RetryCapableBatchWriter will make sure bp1 will only be rewritten before the writing of bp2
I think for now we only need a test in RetryCapableBatchWriterTest for this logic. I added RetryCapableBatchWriterTest.testRetryingKeepChronologicalOrder()